### PR TITLE
feat: Added support for GCP Private Service Connect

### PR DESCRIPTION
**Pull request by amanpruthi**
This Pr has been tested for a upgrade scenario as well as a fresh setup scenario. before this we need to publish helm chart.
PR: https://github.com/wandb/helm-charts/pull/102

Required provider upgrade 

terraform init -upgrade
terraform plan/apply

### CODE REVIEW

File: modules/private_link/main.tf [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  metadata {
+    name = var.ingress_name
+  }
+}
+
+locals {
+  lb_name = data.kubernetes_ingress_v1.ingress.metadata[0].annotations != null ? data.kubernetes_ingress_v1.ingress.metadata[0].annotations["ingress.kubernetes.io/forwarding-rule"] : ""
+}
+
+resource "google_compute_service_attachment" "attachment" {
```
Comment (jsbroks) [inline modules/private_link/main.tf:11]:
```suggestion
resource "google_compute_service_attachment" "default" {
```

File: modules/private_link/main.tf [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+resource "google_compute_subnetwork" "subnet" {
+  name          = "${var.namespace}-psc-ilb-subnet"
+  network       = var.network.id
+  purpose       = "PRIVATE_SERVICE_CONNECT"
+  ip_cidr_range = var.psc_subnetwork
+}
+
+# proxy-only subnet
+resource "google_compute_subnetwork" "proxy_subnet" {
```
Comment (jsbroks) [inline modules/private_link/main.tf:36]:
```suggestion
resource "google_compute_subnetwork" "proxy" {
```

File: modules/private_link/main.tf [hunk 3]
```diff
@@ -0,0 +1,66 @@
+data "kubernetes_ingress_v1" "ingress" {
+  metadata {
+    name = var.ingress_name
+  }
+}
+
+locals {
+  lb_name = data.kubernetes_ingress_v1.ingress.metadata[0].annotations != null ? data.kubernetes_ingress_v1.ingress.metadata[0].annotations["ingress.kubernetes.io/forwarding-rule"] : ""
```
Comment (jsbroks) [inline modules/private_link/main.tf:8]: `lb_name` but you are looking at the forward rulename?
Is this value set by the ingress controller when provisioning?

Comment (amanpruthi) [reply modules/private_link/main.tf:8]: This is not being tested ... It will be tested after helm version upgrade to 1.13

Comment (jsbroks) [reply modules/private_link/main.tf:8]: Okay I approved the helm PR. I would perf not to use data type and pass in the name explicitly

**Comment by jsbroks** [comment]
We should probably add a comment saying that this requires the min version of that helm chart in order to work. So once you merge the helm could you comment back on the chart version

File: modules/private_link/main.tf (hunk 3 above)
Comment (jsbroks) [reply modules/private_link/main.tf:8]: So lets make sure to test it with the new helm chart before we merge it

Comment (amanpruthi) [reply modules/private_link/main.tf]: @jsbroks  ingress doesn't have any annotation that allows you specify the loadbalancer name that will be created by the gce controller .
And we need the name to link the internal load balancer to the private service .
Checked the code also there is no such supported annotation https://github.com/kubernetes/ingress-gce/blob/master/pkg/annotations/ingress.go

File: main.tf [hunk 4]
```diff
@@ -338,3 +361,54 @@ module "wandb" {
...
+  command              = "sleep 450; date +%s"
+  command_when_destroy = "sleep 450"
+  trigger              = timestamp()
+  working_dir          = "/tmp"
+
+  depends = [
+    module.wandb
+  ]
+}
+
```
Comment (jsbroks) [inline main.tf:392]: Why do we need this?

File: main.tf [hunk 5]
```diff
@@ -338,3 +361,54 @@ module "wandb" {
...
+data "google_compute_forwarding_rules" "my_forwarding_rules" {
```
Comment (jsbroks) [inline main.tf:393]: Should probably rename this

File: variables.tf [hunk 6]
```diff
@@ -254,6 +254,44 @@ variable "parquet_wandb_env" {
...
+variable "public_access" {
+  type        = bool
+  description = "Whether to create a public endpoint for wandb access."
+  default = true
+}
+
+variable "allowed_projects" {
+  type = map(number)
+  default = {
+    # "482878270665" = 4
```
Comment (jsbroks) [inline variables.tf:278]: Is it possible to make this project name?

Comment (amanpruthi) [reply variables.tf:278]: Project name not supported we can only pass porject ID or project number, changed variable name now

variable "allowed_project_names" {
  type = map(number)
  default = {
    # "project_ID" = 4 
  }
  description = "A map of allowed project name where each key is a project number or ID and the value(4) is the connection limit."
}

File: main.tf (hunk 5 above)
Comment (amanpruthi) [reply main.tf:393]: data "google_compute_forwarding_rules" "all" {}

File: main.tf (hunk 4 above)
Comment (amanpruthi) [reply main.tf:392]: To create a private link, we need to target an internal load balancer (forwarding rule). After deploying the secondary ingress via the Helm chart, it takes approximately 3 to 4 minutes for the internal load balancer to be created on GCP. Due to this delay, it's necessary to incorporate a sleep module with a duration of at least 450 seconds. This ensures that the private link resource does not fail during the provisioning process.

**Review by jsbroks — APPROVED** [review]
Approving but I have not tested. I just want to make sure this doesn't have any creating changes and it is forward and backfowards compatibility. Could you also indicuate which version of the helm chart this needs to be on

**Comment by amanpruthi** [comment]
it required operator wandb helm chart version 0.13.0

**Comment by jsbroks** [comment]
This PR is included in version 3.6.0 :tada:

**Comment by zacharyblasczyk** [comment]
Technically this was a breaking change as you can't downgrade the version of the provider
: 
Screen Shot 2024-07-15 at 12 40 52 PM
ersion:

**Pull request merged**